plutus-ir/src/Language/PlutusIR/Parser.hs

```diff
@@ -0,0 +1,344 @@
+
+lexeme :: Parser a -> Parser a
+lexeme = Lex.lexeme whitespace
+
+symbol :: String -> Parser String
+symbol = Lex.symbol whitespace
+
+lparen :: Parser String
+lparen = symbol "("
+rparen :: Parser String
+rparen = symbol ")"
+
+lbracket :: Parser String
+lbracket = symbol "["
+rbracket :: Parser String
+rbracket = symbol "]"
+
+lbrace :: Parser String
+lbrace = symbol "{"
+rbrace :: Parser String
+rbrace = symbol "}"
+
+inParens :: Parser a -> Parser a
+inParens = between lparen rparen
+
+inBrackets :: Parser a -> Parser a
+inBrackets = between lbracket rbracket
+
+inBraces :: Parser a -> Parser a
+inBraces = between lbrace rbrace
+
+builtins :: [(String, PLC.BuiltinName)]
+builtins =
+    [ ("addInteger", PLC.AddInteger)
+    , ("subtractInteger", PLC.SubtractInteger)
+    , ("multiplyInteger", PLC.MultiplyInteger)
+    , ("divideInteger", PLC.DivideInteger)
+    , ("quotientInteger", PLC.QuotientInteger)
+    , ("modInteger", PLC.ModInteger)
+    , ("remainderInteger", PLC.RemainderInteger)
+    , ("lessThanInteger", PLC.LessThanInteger)
+    , ("lessThanEqualsInteger", PLC.LessThanEqInteger)
+    , ("greaterThanInteger", PLC.GreaterThanInteger)
+    , ("greaterThanEqualsInteger", PLC.GreaterThanEqInteger)
+    , ("equalsInteger", PLC.EqInteger)
+    , ("resizeInteger", PLC.ResizeInteger)
+    , ("intToByteString", PLC.IntToByteString)
+    , ("concatenate", PLC.Concatenate)
+    , ("takeByteString", PLC.TakeByteString)
+    , ("dropByteString", PLC.DropByteString)
+    , ("equalsByteString", PLC.EqByteString)
+    , ("resizeByteString", PLC.ResizeByteString)
+    , ("sha2_256", PLC.SHA2)
+    , ("sha3_256", PLC.SHA3)
+    , ("verifySignature", PLC.VerifySignature)
+    , ("equalsByteString", PLC.EqByteString)
+    , ("txhash", PLC.TxHash)
+    , ("blocknum", PLC.BlockNum)
+    , ("sizeOfInteger", PLC.SizeOfInteger)
+    ]
+
+reservedWords :: [String]
+reservedWords =
+    map fst builtins ++
+    [ "abs"
+    , "lam"
+    , "ifix"
+    , "fun"
+    , "all"
+    , "bytestring"
+    , "integer"
+    , "size"
+    , "type"
+    , "program"
+    , "con"
+    , "iwrap"
+    , "builtin"
+    , "unwrap"
+    , "error"
+    -- pir-exclusive reserved words
+    , "vardecl"
+    , "typedecl"
+    , "let"
+    , "nonrec"
+    , "rec"
+    , "termbind"
+    , "typebind"
+    , "datatypebind"
+    , "datatype"
+    ]
+
+-- Better to use oneOf?
+identifierChar :: Parser Char
+identifierChar = alphaNumChar <|> char '_' <|> char '\''
+
+reservedWord :: String -> Parser SourcePos
+reservedWord w = lexeme $ do
+    p <- getPosition
+    void $ string w
+    return p
+
+builtinName :: Parser PLC.BuiltinName
+builtinName = lexeme $ choice $ map parseBuiltinName builtins
+    where parseBuiltinName (n, builtin) = try $ string n >> pure builtin
+
+-- TODO: backtracking is probably not needed, remove
+name :: Parser (Name SourcePos)
+name = lexeme $ try $ do
+    pos <- getPosition
+    word <- (:) <$> letterChar <*> many identifierChar
```

Comment:
Never use `many` if possible. Docs of `attoparsec` that apply to `megaparsec` as well:

> Use the Text-oriented parsers whenever possible, e.g. `takeWhile1` instead of `many1 anyChar`. There is about a factor of 100 difference in performance between the two kinds of parser.